std/experimental/checkedint.d

```diff
@@ -0,0 +1,3035 @@
+
+///
+unittest
+{
+    int[] concatAndAdd(int[] a, int[] b, int offset)
+    {
+        // Aborts on overflow on size computation
+        auto r = new int[(checked(a.length) + b.length).get];
+        // Aborts on overflow on element computation
+        foreach (i; 0 .. a.length)
+            r[i] = (a[i] + checked(offset)).get;
+        foreach (i; 0 .. b.length)
+            r[i + a.length] = (b[i] + checked(offset)).get;
+        return r;
+    }
+    assert(concatAndAdd([1, 2, 3], [4, 5], -1) == [0, 1, 2, 3, 4]);
+}
+
+/**
+Checked integral type wraps an integral `T` and customizes its behavior with the
+help of a `Hook` type. The type wrapped must be one of the predefined integrals
+(unqualified), or another instance of `Checked`.
+*/
+struct Checked(T, Hook = Abort)
+if (isIntegral!T || is(T == Checked!(U, H), U, H))
+{
+    import std.algorithm.comparison : among;
+    import std.traits : hasMember;
+    import std.experimental.allocator.common : stateSize;
+
+    /**
+    The type of the integral subject to checking.
+    */
+    alias Representation = T;
+
+    // state {
+    static if (hasMember!(Hook, "defaultValue"))
+        private T payload = Hook.defaultValue!T;
+    else
+        private T payload;
+    /**
+    `hook` is a member variable if it has state, or an alias for `Hook`
+    otherwise.
+    */
+    static if (stateSize!Hook > 0) Hook hook;
+    else alias hook = Hook;
+    // } state
+
+    // get
+    /**
+    Returns a copy of the underlying value.
+    */
+    auto get() inout { return payload; }
+    ///
+    unittest
+    {
+        auto x = checked(ubyte(42));
+        static assert(is(typeof(x.get()) == ubyte));
+        assert(x.get == 42);
+        const y = checked(ubyte(42));
+        static assert(is(typeof(y.get()) == const ubyte));
+        assert(y.get == 42);
+    }
+
+    /**
+    Defines the minimum and maximum. These values are hookable by defining
+    `Hook.min` and/or `Hook.max`.
+    */
+    static if (hasMember!(Hook, "min"))
+    {
+        enum Checked!(T, Hook) min = Checked!(T, Hook)(Hook.min!T);
+        ///
+        unittest
+        {
+            assert(Checked!short.min == -32768);
+            assert(Checked!(short, WithNaN).min == -32767);
+            assert(Checked!(uint, WithNaN).max == uint.max - 1);
+        }
+    }
+    else
+        enum Checked!(T, Hook) min = Checked(T.min);
+    /// ditto
+    static if (hasMember!(Hook, "max"))
+        enum Checked!(T, Hook) max = Checked(Hook.max!T);
+    else
+        enum Checked!(T, Hook) max = Checked(T.max);
+
+    /**
+    Constructor taking a value properly convertible to the underlying type. `U`
+    may be either an integral that can be converted to `T` without a loss, or
+    another `Checked` instance whose representation may be in turn converted to
+    `T` without a loss.
+    */
+    this(U)(U rhs)
+    if (valueConvertible!(U, T) ||
+        !isIntegral!T && is(typeof(T(rhs))) ||
```

Comment:
`isIntegral!T` is always true due to the constraint on the `Checked` struct declaration.